presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/FunctionCallProvider.java

```diff
@@ -0,0 +1,113 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.assertions;
+
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.WindowFrame;
+import com.google.common.base.Joiner;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.toSymbolReferences;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+class FunctionCallProvider
+    implements ExpectedValueProvider<FunctionCall>
+{
+    private boolean isWindowFunction;
+    private final QualifiedName name;
+    private final Optional<WindowFrame> frame;
+    private final boolean distinct;
+    private final List<PlanDslSymbol> args;
+
+    private FunctionCallProvider(boolean isWindowFunction, QualifiedName name, Optional<WindowFrame> frame, boolean distinct, List<PlanDslSymbol> args)
+    {
+        this.isWindowFunction = isWindowFunction;
+        this.name = requireNonNull(name, "name is null");
+        this.frame = requireNonNull(frame, "frame is null");
+        this.distinct = distinct;
+        this.args = requireNonNull(args, "args is null");
+    }
+
+    FunctionCallProvider(QualifiedName name, Optional<WindowFrame> frame, boolean distinct, List<PlanDslSymbol> args)
+    {
+        this(true, name, frame, distinct, args);
+    }
+
+    FunctionCallProvider(QualifiedName name, boolean distinct, List<PlanDslSymbol> args)
+    {
+        this(false, name, Optional.empty(), distinct, args);
+    }
+
+    FunctionCallProvider(QualifiedName name, List<PlanDslSymbol> args)
+    {
+        this(false, name, Optional.empty(), false, args);
+    }
+
+    @Override
+    public String toString()
+    {
+        return format("%s%s (%s) %s", distinct ? "DISTINCT" : "", name, Joiner.on(", ").join(args), frame.isPresent() ? frame.get().toString() : "");
+    }
+
+    public FunctionCall getExpectedValue(ExpressionAliases aliases)
+    {
+        List<Expression> symbolReferences = toSymbolReferences(args, aliases);
+        if (isWindowFunction) {
+            return new ExpectedWindowFunctionCall(symbolReferences);
+        }
+
+        return new FunctionCall(name, symbolReferences);
+    }
+
+    private class ExpectedWindowFunctionCall extends FunctionCall
+    {
+        private ExpectedWindowFunctionCall(List<Expression> args)
+        {
+            super(name, distinct, args);
+        }
+
+        @Override
+        public boolean equals(Object object)
+        {
+            if (this == object) {
+                return true;
+            }
+
+            if (object == null || !(object instanceof FunctionCall)) {
```

Comment:
I've switched to instanceof where the type I'm testing is different from the type of this and getClass where they're the same.